File: arrays_and_hashing/0128_longest_consecutive_sequence.py
# https://leetcode.com/problems/longest-consecutive-sequence/

class Solution:
    def longestConsecutive(self, nums: list[int]) -> int:
        num_set = set(nums)

        longest_subsequence_length = 0

        for num in nums:
            if num - 1 not in num_set:
                current_subsequence_length = 1

                while (num + current_subsequence_length) in num_set:
                    current_subsequence_length += 1

                longest_subsequence_length = max(
                    longest_subsequence_length,
                    current_subsequence_length
                )

        return longest_subsequence_length

File: arrays_and_hashing/0128_longest_consecutive_sequence.cpp
// https://leetcode.com/problems/longest-consecutive-sequence/

#include <unordered_set>
#include <vector>

class Solution {
public:
    int longestConsecutive(std::vector<int>& nums) {
        std::unordered_set<int> uset;

        for (const int& num: nums) {
            uset.insert(num);
        }

        int longestSubsequenceLength = 0;

        for (const int& num: nums) {
            if (uset.find(num - 1) == uset.end()) {
                int currentSequenceLength = 1;

                while (uset.find(num + currentSequenceLength) != uset.end()) {
                    currentSequenceLength++;
                }

                longestSubsequenceLength = std::max(
                    longestSubsequenceLength,
                    currentSequenceLength
                );
            }
        }

        return longestSubsequenceLength;
    }
};
